pkg/net/libp2p/transport.go

```diff
@@ -0,0 +1,294 @@
+	// TODO: Force the termination of incoming connections too and wait here until that happens.
+}
+
+func (tr *Transport) Connect(nodes map[t.NodeID]t.NodeAddress) {
+	tr.connectionsLock.Lock()
+	defer tr.connectionsLock.Unlock()
+
+	for nodeID, addr := range nodes {
+		// For each given node
+
+		if _, ok := tr.connections[nodeID]; !ok {
+			// If connection doesn't yet exist
+
+			// Create a new connection.
+			// This is a non-blocking call - all I/O will be done by the connection in the background.
+			// A connection to self needs a special treatment, as libp2p does not allow to dial oneself.
+			var conn connection
+			var err error
+			if nodeID == tr.ownID {
+				conn, err = newSelfConnection(tr.params, tr.ownID, addr, tr.incomingMessages)
+			} else {
+				conn, err = newRemoteConnection(
+					tr.params,
+					tr.ownID,
+					addr,
+					tr.host,
+					logging.Decorate(tr.logger, "SND: ", "dest", nodeID),
+				)
+			}
+
+			// Complain and ignore node if connection cannot be created.
+			// Note that this is different from not (yet) establishing an underlying network connection,
+			// which does not result in an error and will be retried in the background.
+			if err != nil {
+				tr.logger.Log(logging.LevelError, "Cannot create new connection, ignoring destination.",
+					"err", err, "nodeID", nodeID, "addr", addr)
+			} else {
+				tr.connections[nodeID] = conn
+				tr.nodeIDs[conn.PeerID()] = nodeID
+			}
+		}
+	}
+}
+
+func (tr *Transport) Send(dest t.NodeID, msg *messagepb.Message) error {
+	conn, err := tr.getConnection(dest)
+	if err != nil {
+		return err
+	}
+	return conn.Send(msg)
+}
+
+func (tr *Transport) WaitFor(n int) {
+
+	tr.connectionsLock.RLock()
+
+	// We store the cancel functions returned by all calls to conn.Wait().
+	// When done waiting (for whatever reason), call all those functions
+	// to unblock the goroutines that are still waiting.
+	cancelFuncs := make([]func(), 0, len(tr.connections))
+	defer func() {
+		for _, cancelFunc := range cancelFuncs {
+			cancelFunc()
+		}
+	}()
+
+	// Invoke Wait() on each connection and have the value it outputs written to the aggregator channel.
+	counterChan := make(chan error) // All the outputs produced by conn.Wait() calls are aggregated here.
+	for _, conn := range tr.connections {
+		errChan, cancelFunc := conn.Wait()
+		cancelFuncs = append(cancelFuncs, cancelFunc)
+		go func() {
+			counterChan <- <-errChan // Read from one channel and directly write the value to another (the aggregator).
+		}()
+	}
+
+	tr.connectionsLock.RUnlock()
+
+	// Read values from the counter channel (aggregator) until enough successful
+	// connections (nil error) have been reached or an error occurs.
+	numConnected := 0
+	for numConnected < n {
+		select {
+		case err := <-counterChan:
+			if err != nil {
+				return // TODO: Return an error here (adaptation of the Transport interface needed)
```

Comment:
Should we address it outside of this PR? If not, let's create an issue to not forget about this

Reply:
Added https://github.com/filecoin-project/mir/issues/339